### PR TITLE
add cassandra client pool metrics

### PR DESCRIPTION
**Pull request by jnelson15**
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/1380)

### CODE REVIEW

**Comment by jnelson15** [comment]
One part I don't like about this is that it registers a new metric name for each host. While we want to know these metrics on a per-host basis, I imagine it'll make scraping these easier if there's a standard naming. Any thoughts on how we can achieve both goals? Maybe there's a way to register as same name but tag with host name?

File: atlasdb-api/build.gradle [hunk 1]
```diff
@@ -11,6 +11,7 @@ dependencies {
   compile group: 'com.fasterxml.jackson.core', name: 'jackson-annotations'
   compile group: 'com.fasterxml.jackson.core', name: 'jackson-databind'
   compile group: 'javax.validation', name: 'validation-api'
+  compile group: 'com.codahale.metrics', name: 'metrics-core'
```
Comment (schlosna) [inline atlasdb-api/build.gradle]: Will this conflict with the io.dropwizard 3.1.2 version?

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPool.java [hunk 2]
```diff
@@ -102,6 +103,7 @@
     Map<InetSocketAddress, CassandraClientPoolingContainer> currentPools = Maps.newConcurrentMap();
     final CassandraKeyValueServiceConfig config;
     final ScheduledThreadPoolExecutor refreshDaemon;
+    private final Optional<MetricRegistry> metricRegistry;
```
Comment (schlosna) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPool.java]: Would prefer we don't use optional for field, and in fact would like to require metrics

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolingContainer.java [hunk 3]
```diff
@@ -241,7 +252,44 @@ public String toString() {
...
+                });
+    }
 
-        return new GenericObjectPool<>(cassandraClientFactory, poolConfig);
+    private void registerMetric(GenericObjectPool<Client> pool, String metricPrefix, String metricName, Gauge gauge) {
+        if (!metricRegistry.isPresent()) {
+            return;
+        }
+        String fullyQualifiedMetricName = MetricRegistry.name(
+                CassandraClientPoolingContainer.class, metricPrefix + metricName);
```
Comment (schlosna) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolingContainer.java]: Can avoid the string concatenation here and fall back on varargs method, though should make sure we don't append dot on prefix

**Comment by schlosna** [comment]
Started looking through, will dig in more tomorrow

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPool.java (hunk 2 above)
Comment (jnelson15) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPool.java]: How would that work? Should we break the API and force the clients to create a `MetricRegistry`, or would we have some type of default?

Comment (schlosna) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPool.java]: Probably need to hook in a global MetricRegistry per https://github.com/palantir/atlasdb/pull/1356

File: atlasdb-api/build.gradle (hunk 1 above)
Comment (SerialVelocity) [inline atlasdb-api/build.gradle]: Yes, it does. I had this problem recently due to conflicting classpaths.

**Comment by jnelson15** [comment]
Cool will wait for https://github.com/palantir/atlasdb/pull/1356 to merge and then will modify this to use the global `AtlasDbRegistry` metrics.

**Comment by jnelson15** [comment]
In the meantime, @tpetracca were there any other metrics that you think would be useful given your recent Cassandra debugging? Right now I'm adding mean idle time, mean active time, mean borrow wait time, proportion of connections killed by evictor, and proportion of connections killed by borrower.

**Comment by tpetracca** [comment]
I'd love to be able to understand how much thrashing we do. Like within some granularity I want to see the size of the cache, with the hopes of trying to get a feeling for how "bursty" my clients are and if we're using a lot of resources continually spinning up new clients (because they always clear from idle time after a burst usage).

**Comment by SerialVelocity** [comment]
@jnelson15 #1429 has now merged!

**Comment by jnelson15** [comment]
Awesome, thanks! Will get this one updated soon.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPool.java [hunk 4]
```diff
@@ -537,7 +560,20 @@ private void refreshTokenRanges() {
     private <V, K extends Exception> V runOnHost(InetSocketAddress specifiedHost,
                                                  FunctionCheckedException<Cassandra.Client, V, K> fn) throws K {
         CassandraClientPoolingContainer hostPool = currentPools.get(specifiedHost);
-        return hostPool.runWithPooledResource(fn);
+        totalRequests.getAndIncrement();
```
Comment (hsaraogi) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPool.java]: Can we have requests metric per container too? That will involve currentPools.forEach((address, cassandraClientPoolingContainer) ->
               registerMetric(..))

Comment (jnelson15) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPool.java]: In terms of request/exception rates? It has some info on pool stats in the containers at the moment but can add more metrics if you want that.

Comment (jnelson15) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPool.java]: Oh I see what you mean. Like in the CassandraClientPool, track requests per container as well as aggregate ones?

Comment (hsaraogi) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPool.java]: Yes, exactly.

**Comment by hsaraogi** [comment]
It would be good to have a test that ensure the right metrics are being collected, if thats not too much effort.

**Comment by jnelson15** [comment]
Roger

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPool.java [hunk 5]
```diff
@@ -138,6 +145,33 @@ public String toString() {
         }
     }
 
+    private static class RequestMetrics {
+        final AtomicLong totalRequests = new AtomicLong(0);
+        final AtomicLong totalRequestExceptions = new AtomicLong(0);
+        final AtomicLong totalRequestConnectionExceptions = new AtomicLong(0);
```
Comment (jnelson15) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPool.java]: Fixing

**Pull request merged**